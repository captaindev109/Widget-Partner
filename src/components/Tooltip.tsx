import Card from '@/generics/Card';
import Text from '@/generics/Text';
import { device } from '@/utils/constants';
import React from 'react';
import styled, { DefaultTheme } from 'styled-components';
import { CreateSVGComponent } from './SVGIcons/Index';

export interface TooltipProps {
  size: 'small' | 'medium' | 'large';
  text: string;
  direction: 'top' | 'bottom' | 'left' | 'right';
}

interface TooltipStylingProps {
  theme: DefaultTheme;
  $direction: 'top' | 'bottom' | 'left' | 'right';
}

const SIZE = {
  small: '15px',
  medium: '20px',
  large: '25px',
};

const getArrowPosition = ({ $direction, theme }: TooltipStylingProps) => {
  switch ($direction) {
    case 'top':
      return `
        border-color: ${theme.color.gray900} transparent transparent transparent;
        top: 100%;
        left: 50%;
        margin-left: -${theme.size.margin.m5};
      `;
    case 'bottom':
      return `
        border-color: transparent transparent ${theme.color.gray900} transparent;
        bottom: 100%;
        left: 50%;
        margin-left: -${theme.size.margin.m5};
      `;
    case 'left':
      return `
        border-color: transparent transparent transparent ${theme.color.gray900};
        top: 50%;
        left: 100%;
        margin-top: -${theme.size.margin.m5};
      `;
    case 'right':
      return `
        border-color: transparent ${theme.color.gray900} transparent transparent;
        top: 50%;
        right: 100%;
        margin-top: -${theme.size.margin.m5};
      `;
  }
};

const getBoxPosition = ({ $direction }: TooltipStylingProps) => {
  switch ($direction) {
    case 'top':
      return `
        bottom: calc(100% + 10px);
        left: 50%;
        transform: translateX(-50%);
      `;
    case 'bottom':
      return `
        top: calc(100% + 10px);
        left: 50%;
        transform: translateX(-50%);
      `;
    case 'left':
      return `
        right: calc(100% + 10px);
        top: 50%;
        transform: translateY(-50%);
      `;
    case 'right':
      return `
        left: calc(100% + 10px);
        top: 50%;
        transform: translateY(-50%);
      `;
  }
};

const TooltipText = styled(Card)<TooltipStylingProps>`
  z-index: 1;
  position: absolute;
  visibility: hidden;
  width: max-content;

  max-width: 70vw;

  background-color: ${({ theme }) => theme.color.gray900};
  border-radius: ${({ theme }) => theme.size.borderRadius.br4};
  padding: 12px 16px;

  // Box position
  ${getBoxPosition};

  @media ${device.tablet} {
    max-width: 350px;
  }

  // Transition the opacity
  opacity: 0;
  transition: all 0.3s ease-in;

  // Arrow
  &:after {
    content: '';
    position: absolute;
    border-width: 5px;
    border-style: solid;
    ${getArrowPosition};
    transition: border 0.3s ease-in;
  }
`;

const TooltipIcon = styled(Card)`
  display: flex;

  // When hover over the icon, show the text
  &:hover + ${TooltipText} {
    visibility: visible;
    opacity: 1;
  }
`;

const TooltipCard = styled(Card)`
  position: relative;
  display: inline-block;
`;

function Tooltip({ size, text, direction }: TooltipProps) {
  const icon = CreateSVGComponent('Information', { width: SIZE[size], height: SIZE[size] });
  return (
    <TooltipCard>
      <TooltipIcon>{icon}</TooltipIcon>
      <TooltipText $direction={direction}>
        <Text variant="Body3" color="gray100">
          {text}
        </Text>
      </TooltipText>
    </TooltipCard>
  );
}

Tooltip.defaultProps = {
  size: 'medium',
  text: 'Tooltip',
  direction: 'top',
};

export default Tooltip;
