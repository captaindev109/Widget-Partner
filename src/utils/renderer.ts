import React from 'react';
import { ConfigProps, Index, EachComponentObjectChildren, BasicAny } from '@/types/types';
import { UIComponents } from './components';

const recursivelyCreateReactComponents = (
  componentToCreate: BasicAny,
  configuration: ConfigProps,
  children: BasicAny
) => {
  if (typeof componentToCreate !== 'undefined') {
    return React.createElement(
      componentToCreate,
      configuration,
      children &&
        (typeof children === 'string'
          ? children
          : children.map((component: EachComponentObjectChildren, index: Index) => viewRenderer(component, index)))
    );
  }
  null;
};

const viewRenderer: BasicAny = (config: ConfigProps, index: Index) => {
  const children = config.children;
  const newConfig = { ...config, key: index };
  const LookForComponentInUILibrary = UIComponents[config.component];
  return recursivelyCreateReactComponents(LookForComponentInUILibrary, newConfig, children);
};

export default viewRenderer;
