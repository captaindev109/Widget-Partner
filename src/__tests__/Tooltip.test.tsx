import { render } from '@testing-library/react';
import renderer from 'react-test-renderer';
import { FC } from 'react';
import { ThemeProvider } from 'styled-components';
import defaultTheme from '@/theme/theme';
import Tooltip from '@/components/Tooltip';

const ToolTipComponent: FC = () => {
  return (
    <ThemeProvider theme={defaultTheme}>
      <Tooltip size="small" direction="left" text="Left tooltip example" />
    </ThemeProvider>
  );
};

/** Render tests */

describe('Tooltip render tests', () => {
  test('to see if the component render without crashing', () => {
    render(<ToolTipComponent />);
  });
});

/** Snapshot tests */

describe('Snapshot test', () => {
  it('Snapshot test should pass', () => {
    const snap = renderer.create(<ToolTipComponent />).toJSON();
    expect(snap).toMatchSnapshot();
  });
});
