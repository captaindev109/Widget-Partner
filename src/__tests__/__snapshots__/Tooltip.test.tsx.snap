// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Snapshot test Snapshot test should pass 1`] = `
<div
  className="Card-sc-fb64gk-0 Tooltip__TooltipCard-sc-xggov6-2 leCJCh hfUwNk"
>
  <div
    className="Card-sc-fb64gk-0 Tooltip__TooltipIcon-sc-xggov6-1 leCJCh cyLRGr"
  >
    <svg
      fill="none"
      height="15px"
      viewBox="0 0 20 20"
      width="15px"
      xmlns="http://www.w3.org/2000/svg"
    >
      <path
        d="M9.25 15.5V8.5H10.75V15.5H9.25Z"
        fill="#121212"
      />
      <path
        d="M10 6.875C10.6904 6.875 11.25 6.31536 11.25 5.625C11.25 4.93464 10.6904 4.375 10 4.375C9.30964 4.375 8.75 4.93464 8.75 5.625C8.75 6.31536 9.30964 6.875 10 6.875Z"
        fill="#121212"
      />
      <path
        clipRule="evenodd"
        d="M10 20C15.5228 20 20 15.5228 20 10C20 4.47715 15.5228 0 10 0C4.47715 0 0 4.47715 0 10C0 15.5228 4.47715 20 10 20ZM10 18.5C14.6944 18.5 18.5 14.6944 18.5 10C18.5 5.30558 14.6944 1.5 10 1.5C5.30558 1.5 1.5 5.30558 1.5 10C1.5 14.6944 5.30558 18.5 10 18.5Z"
        fill="#121212"
        fillRule="evenodd"
      />
    </svg>
  </div>
  <div
    className="Card-sc-fb64gk-0 Tooltip__TooltipText-sc-xggov6-0 leCJCh fDlHgV"
  >
    <p
      className="Text__StyledText-sc-1jfsuh1-0 cJpxlc"
      color="gray100"
    >
      Left tooltip example
    </p>
  </div>
</div>
`;
